### PR TITLE
checking test spec for correctness

### LINKED ISSUE #1164: Same integration test fails with IntegrationTestSpec but passes with manual runCommandLine()/assertSamsEqual() calls
I discovered some strange behavior in our test suite. This test in `PrintReadsSparkIntegrationTest` passes:

    @Test
    public void testReadFiltering() throws IOException {
        final File samWithOneMalformedRead = new File(getTestDataDir(), "print_reads_one_malformed_read.sam");
        final File outBam = createTempFile("print_reads_testReadFiltering", ".bam");

        ArgumentsBuilder args = new ArgumentsBuilder();
        args.add("--" + StandardArgumentDefinitions.INPUT_LONG_NAME);
        args.add(samWithOneMalformedRead.getCanonicalPath());
        args.add("--" + StandardArgumentDefinitions.OUTPUT_LONG_NAME);
        args.add(outBam.getCanonicalPath());

        runCommandLine(args.getArgsArray());
        SamAssertionUtils.assertSamsEqual(outBam, new File(getTestDataDir(), "expected.print_reads_one_malformed_read.bam"));
    }

But if you re-write it to use `IntegrationTestSpec` with the same input, output, and expected file, it fails with `Sort order differs. File 1: nullFile 2: coordinate`:

    @Test
    public void testReadFiltering() throws IOException {
        final File samWithOneMalformedRead = new File(getTestDataDir(), "print_reads_one_malformed_read.sam");
        final File outBam = createTempFile("print_reads_testReadFiltering", ".bam");

        final IntegrationTestSpec spec = new IntegrationTestSpec(
                " --" + StandardArgumentDefinitions.INPUT_LONG_NAME + " " + samWithOneMalformedRead.getCanonicalPath() +
                " --" + StandardArgumentDefinitions.OUTPUT_LONG_NAME + " " + outBam.getCanonicalPath(),
                Arrays.asList(new File(getTestDataDir(), "expected.print_reads_one_malformed_read.bam").getCanonicalPath())
        );

        spec.executeTest("PrintReadsSpark_testReadFiltering", this);
    }

Possibly this is indicative of a bug in `IntegrationTestSpec`?

### PR DESCRIPTION
**Pull request by akiezun**
fixes #1164

the problem was that the test spec was incorrect - no placeholders.
 I fixed the spec and added a check for specs in general and added a test to show that the check does work.

for @droazen

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java [hunk 1]
```diff
@@ -94,17 +92,20 @@ public void executeTest(final String name, CommandLineProgramTest test) throws I
             tmpFiles.add(fl);
         }
 
-        final String args = String.format(getArgs(), tmpFiles.toArray());
+        final String preFormattedArgs = getArgs();
+        final String formattedArgs = String.format(preFormattedArgs, tmpFiles.toArray());
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java]: Won't this line throw an `IllegalFormatException` if the spec is missing a placeholder, meaning that we'll never reach your new user-friendly error below?

File: src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java [hunk 2]
```diff
@@ -94,17 +92,20 @@ public void executeTest(final String name, CommandLineProgramTest test) throws I
...
         if (expectsException()) {
             // this branch handles the case were we are testing that a walker will fail as expected
-            executeTest(name, test, null, null, tmpFiles, args, getExpectedException());
+            executeTest(name, test, null, null, tmpFiles, formattedArgs, getExpectedException());
         } else {
-            List<String> expectedFileNames = new ArrayList<>();
-            expectedFileNames.addAll(expectedFileNames());
+            final List<String> expectedFileNames = new ArrayList<>(expectedFileNames());
+            if (!expectedFileNames.isEmpty() && preFormattedArgs.equals(formattedArgs)){
+                throw new GATKException("Incorrect test specification - you're expecting " + expectedFileNames.size() + " file(s) the specified arguments do not contain the same number of \"%s\" placeholders");
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java]: Add a unit test for `IntegrationTestSpec` proving that you get this exception when the spec is missing a placeholder.

File: src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java]: no, i have a test that shows that it works. see in `PrintReadsSparkIntegrationTest.java`

File: src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java]: done in `PrintReadsSparkIntegrationTest.java`

Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java]: That test should be a unit test for `IntegrationTestSpec`, not embedded in the integration tests for a random tool

File: src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java (hunk 1 above)
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java]: Can you explain why `String.format()` doesn't throw here? According to the docs it should throw an `IllegalFormatException`

Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java]: Never mind -- just tested it and it indeed does not throw in the too-few format specifiers case (but does throw in the too-many specifiers case). Odd...

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java [hunk 3]
```diff
@@ -220,4 +223,30 @@ public void testReadFiltering() throws IOException {
         runCommandLine(args.getArgsArray());
         SamAssertionUtils.assertSamsEqual(outBam, new File(getTestDataDir(), "expected.print_reads_one_malformed_read.bam"));
     }
-}
\ No newline at end of file
+
+    @Test
+    public void testReadFiltering_asIntegrationTestSpec() throws IOException {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java]: Just replace the original version of the test with the (correct) `IntegrationTestSpec` version, and move the unit testing of `IntegrationTestSpec` itself into an `IntegrationTestSpecUnitTest`

**Comment by droazen** [comment]
Back to @akiezun, merge after addressing comments.

**Pull request merged**